src/surroundings_menu.cpp

```diff
@@ -0,0 +1,1160 @@
+                    if( auto_scroll ) {
+                        ImGui::SetScrollHereY();
+                        auto_scroll = false;
+                    }
+                    monster_data.selected_entry = it;
+                }
+                ImGui::SameLine( 0, 0 );
+                ImGui::PopID();
+                const Creature *mon = it->get_selected_entity();
+                const Character &you = get_player_character();
+                const bool inattentive = you.has_trait( trait_INATTENTIVE );
+                bool sees = mon->sees( you ) && !inattentive;
+                cataimgui::draw_colored_text( sees ? "!" : " ", c_yellow );
+
+                ImGui::TableNextColumn();
+                std::string mon_name;
+                if( mon->is_monster() ) {
+                    mon_name = mon->as_monster()->name();
+                } else {
+                    mon_name = mon->disp_name();
+                }
+                nc_color mon_color = mon->basic_symbol_color();
+                cataimgui::draw_colored_text( mon_name, mon_color );
+                if( is_selected ) {
+                    ImGui::SetScrollHereY();
+                }
+
+                ImGui::TableNextColumn();
+                nc_color hp_color = c_white;
+                std::string hp_text;
+
+                if( mon->is_monster() ) {
+                    mon->as_monster()->get_HP_Bar( hp_color, hp_text );
+                } else {
+                    std::tie( hp_text, hp_color ) = get_hp_bar( mon->as_npc()->hp_percentage(), 100 );
+                }
+                const int bar_max_width = 5;
+                const int bar_width = utf8_width( hp_text );
+                if( bar_width < bar_max_width ) {
+                    hp_text += "<color_c_white>";
+                    for( int i = bar_max_width - bar_width; i > 0; i-- ) {
+                        hp_text += ".";
+                    }
+                    hp_text += "</color>";
+                }
+                cataimgui::draw_colored_text( hp_text, hp_color );
+
+                ImGui::TableNextColumn();
+                nc_color att_color;
+                std::string att_text;
+
+                if( inattentive ) {
+                    att_text = _( "Unknown" );
+                    att_color = c_yellow;
+                } else if( mon->is_monster() ) {
+                    const std::pair<std::string, nc_color> att = mon->as_monster()->get_attitude();
+                    att_text = att.first;
+                    att_color = att.second;
+                } else if( mon->is_npc() ) {
+                    att_text = npc_attitude_name( mon->as_npc()->get_attitude() );
+                    att_color = mon->symbol_color();
+                }
+                cataimgui::draw_colored_text( att_text, att_color );
+
+                ImGui::TableNextColumn();
+                std::string dist_text = get_distance_string( tripoint_rel_ms(), *it->get_selected_pos() );
```

Comment:
Is it possible for the distance or HP of the monster to change between frames here? I believe the game is paused while in this view, so they cannot. It would be really great then if you could store the formatted distance string and HP bar on the `map_entity` so that you only have to compute them once.

I‌ assume the same applies to all entity types, so I‌ won't call them out individually

Reply:
Added caching here like for names. It also helps on the way to generifying tab drawing, though there's still work left to do until I can handle creature complexities while still keeping them as more table columns.